File: app/src/main/java/com/arny/lubereckiy/ui/activities/StartActivity.java
package com.arny.lubereckiy.ui.activities;

import android.os.Bundle;
import android.support.v4.view.MenuItemCompat;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.*;
import android.text.InputType;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;
import com.arny.arnylib.utils.ToastMaker;
import com.arny.arnylib.utils.Utility;
import com.arny.lubereckiy.R;
import com.arny.lubereckiy.adapter.ObjectAdapter;
import com.arny.lubereckiy.adapter.ObjectViewHolder;
import com.arny.lubereckiy.common.Local;
import com.arny.lubereckiy.models.Pikobject;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class StartActivity extends AppCompatActivity implements SwipeRefreshLayout.OnRefreshListener {
    private RecyclerView recyclerView;
    private ObjectAdapter adapter;
    private List<Pikobject> original;
    private SwipeRefreshLayout mSwipeRefreshLayout;
    private String srchTxt;
    private TextView tvNoItems;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_home);
        initUI();
        loadPlans();
    }

    private void loadPlans() {
        Local.loadPikObjects()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnSubscribe(disposable -> mSwipeRefreshLayout.setRefreshing(true))
                .subscribe(
                        pikobjects -> {
                            tvNoItems.setVisibility(pikobjects.size() > 0 ? View.GONE : View.VISIBLE);
                            original = pikobjects;
                            initList(pikobjects);
                        },
                        throwable -> {
                            mSwipeRefreshLayout.setRefreshing(false);
                            tvNoItems.setVisibility(View.VISIBLE);
                            ToastMaker.toastError(StartActivity.this, throwable.getMessage());
                        },
                        () -> mSwipeRefreshLayout.setRefreshing(false));
    }

    private void initUI() {
        mSwipeRefreshLayout = (SwipeRefreshLayout) findViewById(R.id.srl_all_list);
        mSwipeRefreshLayout.setOnRefreshListener(this);
        recyclerView = (RecyclerView) findViewById(R.id.rv_all_objects);
        tvNoItems = (TextView) findViewById(R.id.tv_noitems);
        recyclerView.setLayoutManager(new GridLayoutManager(this, 1, OrientationHelper.VERTICAL, false));
        recyclerView.setItemAnimator(new DefaultItemAnimator());
        initListAdapter();
    }

    private void initListAdapter() {
        adapter = new ObjectAdapter(this, R.layout.pikobject_list_item, new ObjectViewHolder.SimpleActionListener() {
            @Override
            public void openMap(int position) {
                Local.showObjectMap(StartActivity.this, adapter.getItem(position));
            }

            @Override
            public void OnItemClickListener(int position, Object Item) {
                Local.viewObjectGenPlan(StartActivity.this, adapter.getItem(position));
            }
        });
        adapter.setFilter((constraint, item) -> Utility.matcher("(?i).*" + constraint.toString() + ".*", item.getName()));
        recyclerView.setAdapter(adapter);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main_menu, menu);
        MenuItem item = menu.findItem(R.id.action_search);
        if (getSupportActionBar() == null) {
            return false;
        }
        SearchView sv = new SearchView(getSupportActionBar().getThemedContext());
        sv.setInputType(InputType.TYPE_CLASS_TEXT);
        MenuItemCompat.setShowAsAction(item, MenuItemCompat.SHOW_AS_ACTION_COLLAPSE_ACTION_VIEW | MenuItemCompat.SHOW_AS_ACTION_IF_ROOM);
        MenuItemCompat.setActionView(item, sv);
        sv.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                adapter.filter(query, count -> {
                    tvNoItems.setVisibility(count > 0 ? View.GONE : View.VISIBLE);
                });
                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                adapter.filter(newText, count -> {
                    tvNoItems.setVisibility(count > 0 ? View.GONE : View.VISIBLE);
                });
                return false;
            }
        });
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        switch (id) {
            case R.id.action_filter_min_price:
//                adapter.sort((o1, o2) -> o1.getMinPrice().compareTo(o2.getMinPrice()));
//                System.out.println(adapter.getItem(0).getName());
                break;
            case R.id.action_filter_default:
//                System.out.println(adapter.getItem(0).getName());
//                initList(original);
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    private void initList(List<Pikobject> data) {
        Local.setAdapterData(adapter, data, recyclerView, false);
    }

    @Override
    public void onRefresh() {
        loadPlans();
    }
}

File: app/src/main/java/com/arny/pik/StartActivity.kt
package com.arny.pik

import android.annotation.SuppressLint
import android.os.Bundle
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.OrientationHelper
import android.support.v7.widget.SearchView
import android.view.Menu
import android.view.MenuItem
import android.view.View
import com.arny.pik.adapter.ObjectsAdapter
import com.arny.pik.data.Local
import com.arny.pik.data.api.API
import com.arny.pik.data.models.Pikobject
import com.arny.pik.utils.ToastMaker
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_home.*

class StartActivity : AppCompatActivity(), SwipeRefreshLayout.OnRefreshListener {
    private var adapter: ObjectsAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_home)
        initUI()
        loadPlans()
    }

    @SuppressLint("CheckResult")
    private fun loadPlans() {
        API.loadPikObjects()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnSubscribe { disposable -> srl_all_list.isRefreshing = true }
                .subscribe({ pikobjects ->
                            srl_all_list.isRefreshing = false
                            tv_noitems.visibility = if (pikobjects.size > 0) View.GONE else View.VISIBLE
                            initList(pikobjects)
                        }, { throwable ->
                            srl_all_list.isRefreshing = false
                            tv_noitems.visibility = View.VISIBLE
                            ToastMaker.toastError(this@StartActivity, throwable.message)
                        })
    }

    private fun initUI() {
        srl_all_list.setOnRefreshListener(this)
        rv_all_objects.layoutManager = GridLayoutManager(this, 1, OrientationHelper.VERTICAL, false)
        rv_all_objects.itemAnimator = DefaultItemAnimator()
        initListAdapter()
    }

    private fun initListAdapter() {
        adapter = ObjectsAdapter(object : ObjectsAdapter.ObjectsListener {
            override fun openMap(position: Int, item: Pikobject) {
                Local.showObjectMap(this@StartActivity, item)
            }

            override fun onItemClick(position: Int, item: Pikobject) {
                Local.viewObjectGenPlan(this@StartActivity, item)
            }
        })
//        adapter!!.setFilter({ constraint, item -> Utility.matcher("(?i).*" + constraint.toString() + ".*", item.getName()) })
        rv_all_objects.adapter = adapter
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater?.inflate(R.menu.main_menu, menu)
        val item = menu?.findItem(R.id.action_search)
        val sv = item?.actionView as? SearchView
        sv?.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
//                adapter!!.filter(query) { count -> tv_noitems.visibility = if (count > 0) View.GONE else View.VISIBLE }
                return false
            }

            override fun onQueryTextChange(newText: String): Boolean {
//                adapter!!.filter(newText) { count -> tv_noitems.visibility = if (count > 0) View.GONE else View.VISIBLE }
                return false
            }
        })
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        when (id) {
            R.id.action_filter_min_price -> {
            }
            R.id.action_filter_default -> {
            }
        }//                adapter.sort((o1, o2) -> o1.getMinPrice().compareTo(o2.getMinPrice()));
        //                System.out.println(adapter.getItem(0).getName());
        //                System.out.println(adapter.getItem(0).getName());
        //                initList(original);
        return super.onOptionsItemSelected(item)
    }

    private fun initList(data: List<Pikobject>) {
        adapter?.addAll(data)
    }

    override fun onRefresh() {
        loadPlans()
    }
}
